autopilot/scoreattach.go

```diff
@@ -0,0 +1,160 @@
+var _ AttachmentHeuristic = (*ScoreAttachment)(nil)
+
+// SetNodeScores is used to set the internal map from NodeIDs to scores.
+func (s *ScoreAttachment) SetNodeScores(newScores map[NodeID]uint32) {
+	s.Lock()
+	defer s.Unlock()
+
+	s.nodeScores = newScores
+}
+
+// NeedMoreChans is a predicate that should return true if, given the passed
+// parameters, and its internal state, more channels should be opened within
+// the channel graph. If the heuristic decides that we do indeed need more
+// channels, then the second argument returned will represent the amount of
+// additional funds to be used towards creating channels.
+//
+// NOTE: This is a part of the AttachmentHeuristic interface.
+func (s *ScoreAttachment) NeedMoreChans(channels []Channel,
+	funds btcutil.Amount) (btcutil.Amount, uint32, bool) {
+
+	// First check the heuristic constraints whether we can open more
+	// channels.
+	fundsAvailable, numAdditionalChans, needMore := s.constraints.moreChans(
+		channels, funds,
+	)
+	if !needMore {
+		return 0, 0, false
+	}
+
+	s.Lock()
+	defer s.Unlock()
+
+	// Check if there are any of the preferred nodes left that we don't
+	// have channels to.
+	candidates := make(map[NodeID]struct{})
+	for node, _ := range s.nodeScores {
+		candidates[node] = struct{}{}
+	}
+
+	for _, c := range channels {
+		delete(candidates, c.Node)
+	}
+
+	// If none of the candidates are free, then this heuristic cannot open
+	// more channels.
+	freeCandidate := len(candidates) > 0
+	if !freeCandidate {
+		return 0, 0, false
+	}
+
+	return fundsAvailable, numAdditionalChans, true
+}
+
+// Select returns a candidate set of attachment directives that should be
+// executed base on the internal set of node scores, and the current state of
+// the channels set.
+//
+// NOTE: This is a part of the AttachmentHeuristic interface.
+func (s *ScoreAttachment) Select(self *btcec.PublicKey, g ChannelGraph,
+	fundsAvailable btcutil.Amount, numNewChans uint32,
+	skipNodes map[NodeID]struct{}) ([]AttachmentDirective, error) {
+
+	var directives []AttachmentDirective
+
+	// Return immediately if we don't have enough funds to meet the minumum
+	// channel size.
+	if fundsAvailable < s.constraints.MinChanSize {
+		return directives, nil
+	}
+
+	selfPubBytes := self.SerializeCompressed()
+
+	s.Lock()
+	defer s.Unlock()
+
+	// Gather all of the preferred nodes that shouldn't be skipped.
+	var prefNodes []NodeID
+	for node, _ := range s.nodeScores {
+		if _, ok := skipNodes[node]; ok {
+			continue
+		}
+		if bytes.Equal(node[:], selfPubBytes) {
+			continue
+		}
+
+		prefNodes = append(prefNodes, node)
+	}
+
+	// Sort them, such that the nodes with the highest score will come
+	// first in the slice.
+	sort.Slice(prefNodes, func(i, j int) bool {
+		return s.nodeScores[prefNodes[i]] > s.nodeScores[prefNodes[j]]
+	})
+
+	// We'll continue our attachment loop until we've exhausted the current
+	// amount of available funds.
+	for _, node := range prefNodes {
```

Comment:
For the external scores, we may want to insert a bit of jitter in here. So for example adding a supplementary score based on the xor distance (or just regular distance) of a node's pubkey to our own.